runtime/src/system_instruction_processor.rs

```diff
@@ -99,8 +100,8 @@ fn assign(
         return Err(InstructionError::MissingRequiredSignature);
     }
 
-    // guard against sysvars being made
-    if sysvar::check_id(&owner) {
+    // guard against sysvars and native loader programs being made
+    if sysvar::check_id(&owner) || native_loader::check_id(&owner) {
```

Comment:
This is sneaky.  For real safety here I think we need a compile-time check to inform us when additional restrictions are being added here so we can gate them.   For example if there's a new sysvar or native_loader added, during the 33%-66% upgrade window, anybody can kill the cluster with a simple `solana transfer`.

Reply:
Would also be nice to reserve an upgrade path for our known ids like BPFLoader2111111111.... BPFLoader31111111....., etc...

Reply:
oops, odd. This is intended to be pushed to #11928...

Reply:
@mvines 

>  For real safety here I think we need a compile-time check to inform us when additional restrictions are being added here so we can gate them.

This is interesting and timely. :) Well, since Rust 1.46, we can do this to some extent thanks to `const fn` improvements: https://blog.rust-lang.org/2020/08/27/Rust-1.46.0.html#const-fn-improvements. Namely, we can digest the ast tokens at the compile-time and create sha256 and compare it against annootated `frozen_code(digest = "fooBarbase58")` as an auto-generated test just like `frozen_abi(...)`.

> For example if there's a new sysvar or native_loader added, during the 33%-66% upgrade window, anybody can kill the cluster with a simple solana transfer.

Well, the attack isn't involved with `SystemInstruction::Transfer`. This is related to `SystemInstruction::Assign`. Anyway, it's very easy to kill cluster without gating... I'll gate it.

Reply:
> Would also be nice to reserve an upgrade path for our known ids like BPFLoader2111111111.... BPFLoader31111111....., etc...

@jackcmay gotcha!